app/validators/digital_object/dynamic_fields_validator.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+# This class provided the logic for validating dynamic fields against a map generated by DynamicFieldMap.
+class DigitalObject::DynamicFieldsValidator < ActiveModel::EachValidator
+  CORE_TERM_FIELDS = ['uri', 'alt_labels', 'pref_label', 'term_type', 'authority', 'term_type'].freeze
+
+  private
+
+    def generate_errors(digital_object, attribute, value, map)
+      errors_for(map, value, attribute).each { |a| digital_object.errors.add(a[0], a[1]) }
+    end
+
+    # Returns errors that should be recorded for any of the fields in the data given.
+    # Errors include, invalid fields appearing in the data given, field values containing the wrong type of data.
+    def errors_for(field_map, data, path = nil)
+      errors = []
+
+      data.each do |field_or_group_key, value|
+        new_path = [path, field_or_group_key].compact.join('.')
+
+        unless field_map.key?(field_or_group_key)
+          errors.append([new_path, "is not a valid field"])
+          next
+        end
+
+        reduced_map = field_map[field_or_group_key]
+
+        case reduced_map[:type]
+        when 'DynamicFieldGroup'
+          unless value.is_a?(Array)
+            errors.append([new_path, "must contain an array"])
+            next
+          end
+
+          value.each_with_index do |v, i|
+            errors.concat errors_for(reduced_map[:children], v, "#{new_path}[#{i}]")
+          end
+        when 'DynamicField'
+          if (e = field_errors(reduced_map, value))
+            errors.concat e.map { |i| [new_path, i] }
+          end
+        end
+      end
+
+      errors
+    end
+
+    # Returns any errors that are present for the field and value combination.
+    #
+    # @return [String] if there are errors
+    # @return false if there are no errors
+    def field_errors(configuration, value)
+      send("errors_for_#{configuration[:field_type]}_field", configuration, value)
+    end
+
+    def errors_for_string_field(_configuration, value)
+      value.is_a?(String) ? false : ['must be a string']
+    end
+
+    def errors_for_textarea_field(configuration, value)
+      errors_for_string_field(configuration, value)
+    end
+
+    def errors_for_date_field(_configuration, value)
+      return ['must be a string'] unless value.is_a?(String)
+      return ['must be in YYYY-MM-DD format'] unless value.match?(/^-?\d{4}(-(0[1-9]|1[0-2])(-(0[1-9]|[1,2][0-9]|3[0,1]))?)?/)
+
+      begin
+        Date.parse(value)
+      rescue
+        return ['invalid date']
```

Comment:
```suggestion
        return ['is an invalid date']
```
Suggesting a phrasing change to match the style of some other error messages.

Reply:
Updated.